### PR TITLE
[3628] Reactify Delete dialog

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/3628

### CODE REVIEW

**Comment by rart** [comment]
Fix PR title

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    items,
+    open,
+    onClose
+  } = props;
+  const classes = deleteDialogStyles({});
+  const [selectedItems, setSelectedItems] = useState([]);
+
+  return (
+    <Dialog onClose={onClose} aria-labelledby="simple-dialog-title" open={open}>
+      {/* TODO: this dialog title is the same as the one used in PublishDialog and can be reused, move to a global component? */}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: Good catch. This is already done in one of the active draft PRs. Once those are merged, we should replace this one with — as well as previously completed dialogs.

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  } = props;
+
+  return (
+    <DeleteDialogUI
+      items={items}
+      open={true}
+      onClose={onClose}
+    />
+  )
+}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: Keep an eye on how this component evolves. If it just literally wraps the UI without any container logic, then it's not worth keeping. I do anticipate this one having the container logic (e.g. pulling stuff from redux).

DeleteDialog & DeleteDialogUI can both be in this file.

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        </Typography>
+      </MuiDialogTitle>
+      <Grid container spacing={3}>
+        <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
+          <DependencySelectionDelete
+            items={items}
+            siteId={'editorial'}
+            onChange={(result) => {
+              setSelectedItems(result);
+              console.log('result', result)
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: x

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        <Typography variant="subtitle1" className={classes.subtitle}>
+          Selected items will be deleted along with their items. Please review dependent items before deleting as these
+          will end-up with broken link references.
+        </Typography>
+      </MuiDialogTitle>
+      <Grid container spacing={3}>
+        <Grid item xs={12} sm={12} md={12} lg={12} xl={12}>
+          <DependencySelectionDelete
+            items={items}
+            siteId={'editorial'}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: hardcoded

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+function DeleteDialog(props: DeleteDialogProps) {
+  const {
+    items,
+    onClose,
+    onSuccess
+  } = props;
+  const [open, setOpen] = React.useState(true);
+  const [selectedItems, setSelectedItems] = useState([]);
+  const [submissionComment, setSubmissionComment] = useState('');
+  const [apiState, setApiState] = useState({
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: use spread state and remove spreading in all set state calls

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    deleteItems(siteId, user.username, submissionComment, data).subscribe(
+      (response) => {
+        setOpen(false);
+        setApiState({ ...apiState, error: false, submitting: false });
+        onSuccess?.(response);
+        onClose?.(response);
+      },
+      (response) => {
+        if (response) {
+          setApiState({ ...apiState, error: true, errorResponse: (response.response) ? response.response : response });
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: no spread

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        }
+      }
+    );
+
+  };
+
+  function handleErrorBack() {
+    setApiState({ ...apiState, error: false, global: false });
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: no spread

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    // call externalClose fn
+    onClose?.();
+  };
+
+  const handleSubmit = () => {
+    const data = {
+      items: selectedItems
+    };
+
+    setApiState({ ...apiState, submitting: true });
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: no spread

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    deleteItems(siteId, user.username, submissionComment, data).subscribe(
+      (response) => {
+        setOpen(false);
+        setApiState({ ...apiState, error: false, submitting: false });
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: spread

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  const [open, setOpen] = React.useState(true);
+  const [selectedItems, setSelectedItems] = useState([]);
+  const [submissionComment, setSubmissionComment] = useState('');
+  const [apiState, setApiState] = useState({
+    error: false,
+    submitting: false,
+    global: false,
+    errorResponse: null
+  });
+  const user = useSelector<GlobalState, GlobalState['user']>(state => state.user);
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: isn’t there a custom hook for this?

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+import MuiDialogTitle from '@material-ui/core/DialogTitle';
+import Typography from '@material-ui/core/Typography';
+import IconButton from '@material-ui/core/IconButton';
+import CloseIcon from '@material-ui/icons/Close';
+import { Item } from '../../../models/Item';
+import makeStyles from '@material-ui/core/styles/makeStyles';
+import createStyles from '@material-ui/core/styles/createStyles';
+import { palette } from '../../../styles/theme';
+import { DependencySelectionDelete } from '../Dependencies/DependencySelection';
+import { TextField } from '@material-ui/core';
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: import from its own file

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+                </form>
+              </div>
+              <div className={classes.dialogActions}>
+                <Button variant="contained" onClick={handleClose} disabled={apiState.submitting}>
+                  <FormattedMessage
+                    id="deleteDialog.cancel"
+                    defaultMessage={`Cancel`}
+                  />
+                </Button>
+                <Button variant="contained" autoFocus onClick={handleSubmit} color="primary"
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: Format

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+                    apiState.submitting ?
+                      (
+                        <CircularProgress
+                          className={classes.btnSpinner}
+                          size={20}
+                        />
+                      ) : (
+                        <FormattedMessage
+                          id="deleteDialog.submit"
+                          defaultMessage={`Delete`}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: format

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+                      className: classes.textField
+                    }}
+                  />
+                </form>
+              </div>
+              <div className={classes.dialogActions}>
+                <Button variant="contained" onClick={handleClose} disabled={apiState.submitting}>
+                  <FormattedMessage
+                    id="deleteDialog.cancel"
+                    defaultMessage={`Cancel`}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: format

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+                  <FormattedMessage
+                    id="deleteDialog.headerSubTitle"
+                    defaultMessage="Selected items will be deleted along with their items. Please review dependent items before deleting as these will end-up with broken link references."
+                  />
+                </Typography>
+              </MuiDialogTitle>
+              <div className={classes.dialogContent}>
+                <DependencySelectionDelete
+                  items={items}
+                  siteId={'editorial'}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: hardcoded

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+  } = props;
+  const classes = deleteDialogStyles({});
+
+  return (
+    <Dialog
+      onClose={onClose}
+      aria-labelledby="simple-dialog-title"
+      open={open}
+      fullWidth={true}
+      maxWidth={'sm'}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: format

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+    onClose
+  } = props;
+  const classes = deleteDialogStyles({});
+
+  return (
+    <Dialog
+      onClose={onClose}
+      aria-labelledby="simple-dialog-title"
+      open={open}
+      fullWidth={true}
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: simplify bool syntax

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+                <DependencySelectionDelete
+                  items={items}
+                  siteId={'editorial'}
+                  onChange={(result) => {
+                    setSelectedItems(result);
+                  }}
+                />
+                <form className={classes.submissionCommentField} noValidate autoComplete="off">
+                  <TextField
+                    id="outlined-multiline-static"
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: use appropriate id and camelcased

File: ui/app/src/services/content.ts [hunk 19]
```diff
@@ -1039,6 +1039,24 @@ export function uploadDataUrl(
   });
 }
 
+export function deleteItems(siteId: string, user: string, submissionComment: string, data): Observable<any> {
```
Comment (rart) [inline ui/app/src/services/content.ts]: data missing type

File: ui/app/src/services/content.ts [hunk 20]
```diff
@@ -1039,6 +1039,24 @@ export function uploadDataUrl(
...
+  return post(
```
Comment (rart) [inline ui/app/src/services/content.ts]: use post json util

**Review by rart — CHANGES_REQUESTED** [review]

**Comment by rart** [comment]
Pull/merge conflict

**Comment by rart** [comment]
DialogTitle component is now merged

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx (hunk 10 above)
Comment (jvega190) [reply ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: @rart there wasn't. I Created one, please let me know if it's ok

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx (hunk 1 above)
Comment (rart) [reply ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: Dialog title is now merged.

File: ui/app/src/components/ComponentMenu.tsx [hunk 21]
```diff
@@ -77,6 +83,7 @@ export default function ComponentMenu(props: ComponentMenuProps) {
       getItem(site, path).subscribe(
         (item) => {
           setPublishDialog({ item });
+          setDeleteDialog({ items: [item] });
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx:89]: Wrap API response in service

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx (hunk 16 above)
Comment (rart) [reply ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: 👆🏻

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+    siteId,
+    onClose
+  } = props;
+  const classes = deleteDialogStyles({});
+  const { formatMessage } = useIntl();
+
+  return (
+    <Dialog
+      onClose={onClose}
+      aria-labelledby="simple-dialog-title"
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: Where is this ID?
Use a semantic/specific title

File: ui/app/src/services/content.ts [hunk 23]
```diff
@@ -1049,6 +1049,21 @@ export function getQuickCreateContentList(siteId: string) {
   )
 }
 
+export function deleteItems(siteId: string, user: string, submissionComment: string, data: AnyObject): Observable<any> {
+  return postJSON(
+    `/studio/api/1/services/api/1/workflow/go-delete.json?site=${siteId}&user=${user}&submissionComment=${submissionComment}`,
+    data
+  ).pipe(
+    map((response: any) => {
+      if (response.response.success) {
```
Comment (rart) [inline ui/app/src/services/content.ts]: Abstract error response in API 2 error response

File: ui/app/src/components/ComponentMenu.tsx [hunk 24]
```diff
@@ -238,9 +251,16 @@ export default function ComponentMenu(props: ComponentMenuProps) {
...
+        <DeleteDialog
+          items={deleteDialog.items}
+          onClose={onCloseDelete}
+        />
+      }
       {
         error &&
-        <ErrorDialog error={error} onClose={onErrorClose} />
+        <ErrorDialog error={error} onClose={onErrorClose}/>
       }
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: Dialog manager

File: ui/app/src/services/content.ts [hunk 25]
```diff
@@ -1049,6 +1049,20 @@ export function getQuickCreateContentList(siteId: string) {
   )
 }
 
+export function deleteItems(siteId: string, user: string, submissionComment: string, data: AnyObject): Observable<any> {
+  return postJSON(
+    `/studio/api/1/services/api/1/workflow/go-deletae.json?site=${siteId}&user=${user}&submissionComment=${submissionComment}`,
+    data
+  ).pipe(
+    map((response: any) => {
+      if (response.response.success) {
```
Comment (rart) [inline ui/app/src/services/content.ts]: When would `!response.success`? What then?

**Review by rart — CHANGES_REQUESTED** [review]
Co-locate DeleteDialog & DeleteDialogUI on same file

File: ui/app/src/components/ComponentMenu.tsx [hunk 26]
```diff
@@ -67,6 +70,10 @@ export default function ComponentMenu(props: ComponentMenuProps) {
     scheduling: null
   });
 
+  const [deleteDialog, setDeleteDialog] = useSpreadState({
+    items: []
+  });
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: Spread state with a single prop doesn't make too much sense.

File: ui/app/src/components/ComponentMenu.tsx [hunk 27]
```diff
@@ -240,7 +252,7 @@ export default function ComponentMenu(props: ComponentMenuProps) {
       }
       {
         error &&
-        <ErrorDialog error={error} onClose={onErrorClose} />
+        <ErrorDialog error={error} onClose={onErrorClose}/>
       }
```
Comment (rart) [inline ui/app/src/components/ComponentMenu.tsx]: Dispatch to dialog manager

File: ui/app/src/models/GlobalState.ts [hunk 28]
```diff
@@ -112,6 +113,7 @@ export interface GlobalState {
   };
   dialogs: {
     confirm: ConfirmDialogStateProps;
+    deleteDialog: DeleteDialogStateProps;
```
Comment (rart) [inline ui/app/src/models/GlobalState.ts]: Follow the pattern: remove the word dialog

File: ui/app/src/services/dependencies.ts [hunk 29]
```diff
@@ -18,11 +18,21 @@
...
 
 export function fetchDependencies(siteId: string, items: any) {
   return get(`/studio/api/2/dependency/dependencies?siteId=${siteId}&paths=${items}`)
 }
 
+export function fetchDeleteDependencies(siteId: string, paths: string[]) {
+  return get(`/studio/api/2/content/get_delete_package?siteId=${siteId}&paths=${paths}`).pipe(
+    map((response: any) => {
+      return response.response
+    })
```
Comment (rart) [inline ui/app/src/services/dependencies.ts]: pluck

File: ui/app/src/state/reducers/dialogs/index.ts [hunk 30]
```diff
@@ -16,7 +16,9 @@
 
 import { combineReducers } from '@reduxjs/toolkit';
 import confirm from './confirm';
+import deleteDialog from './delete';
 
 export default combineReducers({
-  confirm
+  confirm,
+  deleteDialog
```
Comment (rart) [inline ui/app/src/state/reducers/dialogs/index.ts]: Follow the pattern: remove the word dialog

File: ui/app/src/modules/Content/Dependencies/DependencySelection.tsx [hunk 31]
```diff
@@ -294,94 +305,57 @@ export function DependencySelectionDelete(props: DependencySelectionProps) {
...
+        <div className={classes.bottomSection}>
+          {
+            (resultItems === null) && (
+              <div className="centerCircularProgress">
+                <CenterCircularProgress/>
+                <span className={classes.circularProgressText}>
                   <FormattedMessage
-                    id="deleteDialog.brokenItems"
-                    defaultMessage={` Will have broken references`}
+                    id="deleteDialog.uploadingDepenedents"
```
Comment (rart) [inline ui/app/src/modules/Content/Dependencies/DependencySelection.tsx]: uploading?
Depenedents?

Typos

File: ui/app/src/modules/Content/Dependencies/DependencySelection.tsx [hunk 32]
```diff
@@ -294,94 +305,57 @@ export function DependencySelectionDelete(props: DependencySelectionProps) {
...
+            title={
+              <FormattedMessage
+                id="deleteDialog.dependentItems"
+                defaultMessage={`Dependent Items`}
+              />
+            }
+            subtitle={
+              <FormattedMessage
+                id="deleteDialog.brokenItems"
+                defaultMessage={` Will have broken references`}
```
Comment (rart) [inline ui/app/src/modules/Content/Dependencies/DependencySelection.tsx]: Is the space at the beginning intentional? There be a better way of achieving it even if it is.
Please change all these curly brackets with side ticks to regular double quotes

File: ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+interface DeleteDialogContentUIProps {
+  resource: Resource<DeleteDependencies>;
+  items: Item[];
+  submissionComment: string;
+  setSubmissionComment: Function;
+
+  onSelectionChange?: Function;
+};
+
+function DeleteDialogContentUI(props: DeleteDialogContentUIProps) {
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialogUI.tsx]: Move DeleteDialogUI to DeleteDialog.tsx (i.e. remove this file and move all to the main file).

File: ui/app/src/modules/Content/Delete/DeleteDialog.tsx [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+  const [selectedItems, setSelectedItems] = useState([]);
+
+  const resource = useStateResource<any, any>(
+    deleteDependencies,
+    {
+      shouldResolve: (deleteDependencies) => Boolean(deleteDependencies),
+      shouldReject: () => Boolean(apiState.error),
+      shouldRenew: (source, resource) => resource.complete,
+      resultSelector: () => deleteDependencies,
+      errorSelector: () => apiState.error
```
Comment (rart) [inline ui/app/src/modules/Content/Delete/DeleteDialog.tsx]: Do not use closures. Use the supplied arguments of each checker.

**Review by rart — CHANGES_REQUESTED** [review]
Please fix conflicts

File: ui/app/src/state/reducers/dialogs/index.ts [hunk 35]
```diff
@@ -17,8 +17,10 @@
 import { combineReducers } from '@reduxjs/toolkit';
 import confirm from './confirm';
 import error from './error';
+import deleteDialog from './delete';
```
Comment (rart) [inline ui/app/src/state/reducers/dialogs/index.ts:24]: Follow the pattern: remove “dialog”

File: ui/app/src/services/content.ts [hunk 36]
```diff
@@ -59,9 +58,10 @@ export function getContent(site: string, path: string): Observable<string> {
   );
 }
 
-export function getItem(site: string, path: string): Observable<Item> {
+export function getItem(site: string, path: string): Observable<any> {
```
Comment (rart) [inline ui/app/src/services/content.ts]: Checkout the **file** from your other PR

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/state/reducers/dialogs/delete.ts [hunk 37]
```diff
@@ -0,0 +1,37 @@
...
+export const showDeleteDialog = createAction<Partial<DeleteDialogStateProps>>('SHOW_DELETE_DIALOG');
+
+export const closeDeleteDialog = createAction<StandardAction>('CLOSE_DELETE_DIALOG');
+
+export default createReducer<GlobalState['dialogs']['delete']>(
+  { open: false },
+  {
+    [showDeleteDialog.type]: (state, { payload }) => ({
+      onClose: closeDeleteDialog(),
+      onSuccess: closeDeleteDialog(),
```
Comment (rart) [inline ui/app/src/state/reducers/dialogs/delete.ts]: onSuccess not required to provide default

**Review by rart — CHANGES_REQUESTED** [review]
Consolidate other PR updates to shared files

**Comment by jvega190** [comment]
Checked & ready

**Review by rart — APPROVED** [review]

**Pull request merged**